File: Chain/Chain.js
/**
 * Created by jamesyan on 2017/3/25.
 */
const ERROR = 3;
const DEBUG = 2;
const INFO = 1;

var AbstractLogger = function (level) {
    this.level = level;
    this.write=function(){};
}

AbstractLogger.prototype.setNextLogger = function (nextLogger) {
    this.nextLogger = nextLogger;
}


AbstractLogger.prototype.logMessage = function (level, message) {
    if (this.level <= level) {
        this.write.call(this,message);
    }

    if (this.nextLogger != null) {
        this.nextLogger.logMessage(level, message);
    }
}

var ErrorLogger = function (level) {
    AbstractLogger.call(this, level);
    this.write = function (message) {
        console.log("Error Logger:" + message);
    }
}

ErrorLogger.prototype = Object.create(AbstractLogger.prototype);

var FileLogger = function (level) {
    AbstractLogger.call(this, level);
    this.write = function (message) {
        console.log("File Debug Logger:" + message);
    }
}

FileLogger.prototype = Object.create(AbstractLogger.prototype);

var ConsoleLogger = function (level) {
    AbstractLogger.call(this, level);
    this.write = function (message) {
        console.log("standard Console Logger:" + message);
    }
}

ConsoleLogger.prototype = Object.create(AbstractLogger.prototype);

function chainPatternDemo() {
    var errorLogger = new ErrorLogger(ERROR);
    var fileLogger = new FileLogger(DEBUG);
    var consoleLogger = new ConsoleLogger(INFO);
    errorLogger.setNextLogger(fileLogger);
    fileLogger.setNextLogger(consoleLogger)
    return errorLogger;

}

File: Chain/Chain.go
package main

import "fmt"

const (
	DEBUG = 3
	Error = 2
	INFO = 1
)
type Logger interface {
	write(message string)
	getLevel()int
	getNextLogger() Logger

}

type AbstractLogger struct {
	nextLogger Logger
	level      int
}
func(a *AbstractLogger)setNextLogger(nextLogger Logger){
   a.nextLogger=nextLogger
}

func (a *AbstractLogger)getLevel() int{
	return a.level
}

func (a *AbstractLogger)getNextLogger() Logger{
	return a.nextLogger
}

func LogMessage(a Logger,level int,message string){
	if(a.getLevel()<=level){
		a.write(message)
	}

	if(a.getNextLogger()!=nil){
		LogMessage(a.getNextLogger(),level,message)
	}
}

type ConsoleLogger struct {
	AbstractLogger
}

func(e *ConsoleLogger)init(level int){
	e.level=level;
}


func(e *ConsoleLogger)write(message string){
	fmt.Println("Standard Console::Logger: ",message)
}

type ErrorLogger struct {
	AbstractLogger
}

func(e *ErrorLogger)init(level int){
	e.level=level;
}


func(e *ErrorLogger)write(message string){
    fmt.Println("Error Console::Logger: ",message)
}

type FileLogger struct {
	AbstractLogger
}

func(e *FileLogger)init(level int){
	e.level=level;
}

func(e *FileLogger)write(message string){
	fmt.Println("File::Logger: ",message)
}


func chainPatternDemo() Logger{
	errorLogger:=new(ErrorLogger)
	errorLogger.init(Error)
	fileLogger:=new(FileLogger)
	fileLogger.init(DEBUG)
	consoleLogger:=new(ConsoleLogger)
	consoleLogger.init(INFO)
	errorLogger.setNextLogger(fileLogger)
	fileLogger.setNextLogger(consoleLogger)
	return errorLogger
}

func main() {
	logChain:=chainPatternDemo()
	LogMessage(logChain,INFO,"this is an INFO info");
	LogMessage(logChain,DEBUG,"this is an DEBUG info");
	LogMessage(logChain,Error,"this is an Error info");
}
